trace/trace.go

```diff
@@ -0,0 +1,228 @@
+// Package trace wraps a datastore where all datastore interactions are traced
+// with open telemetry.
+package trace
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	ds "github.com/ipfs/go-datastore"
+	dsq "github.com/ipfs/go-datastore/query"
+	"go.opentelemetry.io/otel/attribute"
+	otel "go.opentelemetry.io/otel/trace"
+)
+
+// New returns a new traced datastore. All datastore interactions are traced.
+func New(ds ds.Datastore, tracer otel.Tracer) *Datastore {
+	return &Datastore{ds: ds, tracer: tracer}
+}
+
+// Datastore is an adapter that traces inner datastore interactions.
+type Datastore struct {
+	ds     ds.Datastore
+	tracer otel.Tracer
+}
+
+var (
+	_ ds.Datastore           = (*Datastore)(nil)
+	_ ds.Batching            = (*Datastore)(nil)
+	_ ds.PersistentDatastore = (*Datastore)(nil)
+	_ ds.TxnDatastore        = (*Datastore)(nil)
+	_ ds.CheckedDatastore    = (*Datastore)(nil)
+	_ ds.ScrubbedDatastore   = (*Datastore)(nil)
+	_ ds.GCDatastore         = (*Datastore)(nil)
+	_ io.Closer              = (*Datastore)(nil)
+)
+
+// Put implements the ds.Datastore interface.
+func (t *Datastore) Put(ctx context.Context, key ds.Key, value []byte) (err error) {
+	ctx, span := t.tracer.Start(ctx, "Put", otel.WithAttributes(attribute.String("key", key.String())))
+	defer span.End()
```

Comment:
We're not tracing errors anywhere. We probably want to call `span.RecordError(err)` for all errors here (it'll ignore nil errors).

Reply:
ah good point, I guess it also makes sense to set the status then:

```
	// RecordError will record err as an exception span event for this span. An
	// additional call to SetStatus is required if the Status of the Span should
	// be set to Error, as this method does not change the Span status. If this
	// span is not being recorded or err is nil then this method does nothing.
```

Reply:
Well, that API is bonkers. But ok.